addons/controllers/cpi/cpiconfig_controller.go

```diff
@@ -0,0 +1,314 @@
+		r.Log.Error(err, "Fail to reconcile CPIConfig")
+		return res, err
+	}
+	return ctrl.Result{}, nil
+}
+
+// ReconcileCPIConfig reconciles CPIConfig with its owner cluster
+func (r *CPIConfigReconciler) ReconcileCPIConfig(ctx context.Context, req ctrl.Request, cpiConfig *cpiv1alpha1.CPIConfig, cluster *clusterapiv1beta1.Cluster) (res ctrl.Result, retErr error) {
+	patchHelper, err := clusterapipatchutil.NewHelper(cpiConfig, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// patch CPIConfig before returning the function
+	defer func() {
+		r.Log.Info("Patching CPIConfig")
+		if err := patchHelper.Patch(ctx, cpiConfig); err != nil {
+			r.Log.Error(err, "Error patching CPIConfig")
+			retErr = err
+		}
+		r.Log.Info("Successfully patched CPIConfig")
+	}()
+
+	if !cpiConfig.GetDeletionTimestamp().IsZero() {
+		return r.ReconcileCPIConfigDelete(ctx, req, cpiConfig, cluster)
+	}
+	res, err = r.ReconcileCPIConfigNormal(ctx, req, cpiConfig, cluster)
+	if err != nil {
+		r.Log.Error(err, "Error reconciling CPIConfig to create/patch data values secret")
+	}
+	r.Log.Info("Successfully reconciled CPIConfig")
+	return res, err
+}
+
+// ReconcileCPIConfigNormal triggers when a CPIConfig is not being deleted
+// it ensures the owner reference of the CPIConfig and generates the data values secret for CPI
+func (r *CPIConfigReconciler) ReconcileCPIConfigNormal(ctx context.Context, _ ctrl.Request,
+	cpiConfig *cpiv1alpha1.CPIConfig, cluster *clusterapiv1beta1.Cluster) (ctrl.Result, error) {
+	// add owner reference to CPIConfig if not already added by TanzuClusterBootstrap Controller
+	ownerReference := metav1.OwnerReference{
+		APIVersion: clusterapiv1beta1.GroupVersion.String(),
+		Kind:       cluster.Kind,
+		Name:       cluster.Name,
+		UID:        cluster.UID,
+	}
+	r.Log.Info("Ensure CPIConfig has the cluster as owner reference")
+	if !clusterapiutil.HasOwnerRef(cpiConfig.OwnerReferences, ownerReference) {
+		r.Log.Info("Adding owner reference to CPIConfig")
+		cpiConfig.OwnerReferences = clusterapiutil.EnsureOwnerRef(cpiConfig.OwnerReferences, ownerReference)
+	}
+	secret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      util.GenerateDataValueSecretName(cluster.Name, constants.CPIAddonName),
+			Namespace: cpiConfig.Namespace,
+		},
+		Type: v1.SecretTypeOpaque,
+	}
+	secret.SetOwnerReferences([]metav1.OwnerReference{ownerReference})
+
+	mutateFn := func() error {
+		secret.Data = make(map[string][]byte)
+		cpiConfigSpec, err := r.mapCPIConfigToDataValues(ctx, cpiConfig, cluster)
+		r.Log.Info("Mutate CPIConfig data values", "data values", cpiConfigSpec)
+		if err != nil {
+			r.Log.Error(err, "Error occurs when mutate CPIConfig data values")
+			return err
+		}
+		yamlBytes, err := yaml.Marshal(cpiConfigSpec)
+		if err != nil {
+			r.Log.Error(err, "Error marshaling CPIConfig to Yaml")
+			return err
+		}
+		dataValueBytes := append([]byte(constants.TKGDataValueFormatString), yamlBytes...)
```

Comment:
We don't need to include `TKGDataValueFormatString` anymore, as we are not going to add overlay support, and it was used for overlay support.

Reply:
removed